openedx/core/djangoapps/course_roles/docs/decisions/0002-data_storage

```diff
@@ -0,0 +1,56 @@
+2. Data Storage
+################
+
+Status
+******
+
+**Provisional** *2023-09-13*
+
+Context
+*******
+
+The requirements of the course roles project indicate that equal priority be given to user understanding of the roles and flexibility of the system. It has been determined that in order to prioritize user understadning of roles and permissions it is necessary that the descriptions be localized.
+
+**Localization:** It is necessary for all text that will be displayed in the LMS to be shown in the appropriate language. The current process for translation requires that all strings displayed to users be in strings in the codebase.
+
+**Flexibility:** It is necessary that course_roles allows for the easy creation of new roles. It is considered important, but not strictly required, that each instance of the Open edX code can have its own roles without needing to fork the repo and maintain separate code. 
+
+Decision
+********
+
+We will be using a combination of data storage in the database and data storage structures within the code.
+We will store the roles, permissions, role permissions (rolepermission table), and role assignments (userrole table) in the database. 
+We will store the role names (for default roles), permission names, and permission descriptions in a data object in the code.
+We will write code that pulls the role name from the database table for any role not found in the data object in the code. This applies to non-default roles that may differ from instance to instance.
```

Comment:
To me, this sounds like there are two sources of truth, and I think there's a simpler approach.

Namely: 

* We will store the roles, permissions, role permissions (rolepermission table), role assignments (userrole table), role names (for default roles), permission names, and permission descriptions in the database.
* "Default" roles and permissions used by all Open edX instances will be created in the database via migrations.
* Before displaying role names, permission names, or permission descriptions to the user we will use `pgettext` to translate them into the user's locale if a translation is available.
* Somewhere in the code we will maintain a list of strings, marked for translation, that contains the names and descriptions of the default roles and permissions. This code will not be directly used at runtime, but only exists to ensure that the names/descriptions are detected and made available for localization. (This could even be done in the data migration files themselves, which would be the most efficient.)